File: examples/reference-implementations/typescript-client/src/messaging.ts
/**
 * OSSA Agent-to-Agent Messaging
 *
 * Client methods for A2A messaging, webhooks, and event streaming.
 */

import { OSSAClient } from './client.js';

export interface A2AMessage {
  id?: string;
  from: {
    publisher: string;
    name: string;
    version?: string;
    instance_id?: string;
  };
  to: {
    publisher: string;
    name: string;
    version?: string;
    instance_id?: string;
  };
  type: 'request' | 'response' | 'event' | 'broadcast';
  capability?: string;
  payload: Record<string, unknown>;
  metadata?: {
    correlation_id?: string;
    reply_to?: string;
    timestamp?: string;
    ttl?: number;
    priority?: 'low' | 'normal' | 'high' | 'critical';
  };
}

export interface SendMessageResponse {
  message_id: string;
  status: 'sent' | 'queued' | 'delivered' | 'failed';
  timestamp: string;
  delivery_info?: {
    delivered_at?: string;
    acknowledged_at?: string;
    retry_count?: number;
  };
}

export interface MessageStatus {
  message_id: string;
  status: 'sent' | 'queued' | 'delivered' | 'acknowledged' | 'failed' | 'expired';
  created_at: string;
  updated_at: string;
  delivery_attempts: number;
  error?: string;
}

export interface WebhookConfig {
  id?: string;
  url: string;
  events: string[];
  filters?: {
    publisher?: string;
    agent_name?: string;
    capability?: string;
  };
  headers?: Record<string, string>;
  retry_config?: {
    max_attempts?: number;
    backoff_strategy?: 'linear' | 'exponential';
  };
  active?: boolean;
}

export interface WebhookResponse {
  id: string;
  url: string;
  events: string[];
  active: boolean;
  created_at: string;
  last_triggered_at?: string;
  delivery_stats: {
    total_deliveries: number;
    successful_deliveries: number;
    failed_deliveries: number;
  };
}

export interface EventSubscription {
  id?: string;
  agent: {
    publisher: string;
    name: string;
  };
  events: string[];
  filters?: Record<string, unknown>;
  delivery_mode: 'webhook' | 'polling' | 'streaming';
  config?: {
    webhook_url?: string;
    batch_size?: number;
    batch_timeout?: number;
  };
}

export interface EventStreamOptions {
  agent?: {
    publisher: string;
    name: string;
  };
  event_types?: string[];
  from_timestamp?: string;
  batch_size?: number;
}

export interface AgentEvent {
  id: string;
  event_type: string;
  agent: {
    publisher: string;
    name: string;
    version: string;
  };
  timestamp: string;
  payload: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

/**
 * Messaging operations client
 */
export class MessagingClient {
  constructor(private client: OSSAClient) {}

  /**
   * Send an A2A message
   */
  async sendMessage(message: A2AMessage): Promise<SendMessageResponse> {
    return this.client.request<SendMessageResponse>({
      method: 'POST',
      path: '/messaging/send',
      body: message,
    });
  }

  /**
   * Send a request and wait for response (synchronous-style)
   */
  async sendRequest(
    message: Omit<A2AMessage, 'type'>,
    timeout: number = 30000
  ): Promise<{
    request_id: string;
    response?: A2AMessage;
    status: 'completed' | 'timeout' | 'failed';
    error?: string;
  }> {
    return this.client.request({
      method: 'POST',
      path: '/messaging/request',
      body: {
        ...message,
        timeout,
      },
    });
  }

  /**
   * Get message status
   */
  async getMessageStatus(messageId: string): Promise<MessageStatus> {
    return this.client.request<MessageStatus>({
      method: 'GET',
      path: `/messaging/messages/${messageId}`,
    });
  }

  /**
   * Broadcast message to multiple agents
   */
  async broadcast(
    message: Omit<A2AMessage, 'to' | 'type'>,
    filters: {
      domain?: string;
      capability?: string;
      tags?: string[];
    }
  ): Promise<{
    broadcast_id: string;
    recipients_count: number;
    status: 'broadcasting';
  }> {
    return this.client.request({
      method: 'POST',
      path: '/messaging/broadcast',
      body: {
        message,
        filters,
      },
    });
  }

  /**
   * Register a webhook
   */
  async registerWebhook(config: WebhookConfig): Promise<WebhookResponse> {
    return this.client.request<WebhookResponse>({
      method: 'POST',
      path: '/messaging/webhooks',
      body: config,
    });
  }

  /**
   * List webhooks
   */
  async listWebhooks(): Promise<{ webhooks: WebhookResponse[] }> {
    return this.client.request({
      method: 'GET',
      path: '/messaging/webhooks',
    });
  }

  /**
   * Get webhook details
   */
  async getWebhook(webhookId: string): Promise<WebhookResponse> {
    return this.client.request<WebhookResponse>({
      method: 'GET',
      path: `/messaging/webhooks/${webhookId}`,
    });
  }

  /**
   * Update webhook configuration
   */
  async updateWebhook(webhookId: string, config: Partial<WebhookConfig>): Promise<WebhookResponse> {
    return this.client.request<WebhookResponse>({
      method: 'PUT',
      path: `/messaging/webhooks/${webhookId}`,
      body: config,
    });
  }

  /**
   * Delete webhook
   */
  async deleteWebhook(webhookId: string): Promise<void> {
    return this.client.request({
      method: 'DELETE',
      path: `/messaging/webhooks/${webhookId}`,
    });
  }

  /**
   * Subscribe to agent events
   */
  async subscribe(subscription: EventSubscription): Promise<{
    subscription_id: string;
    status: 'active';
    created_at: string;
  }> {
    return this.client.request({
      method: 'POST',
      path: '/messaging/subscriptions',
      body: subscription,
    });
  }

  /**
   * List event subscriptions
   */
  async listSubscriptions(): Promise<{
    subscriptions: Array<{
      id: string;
      agent: { publisher: string; name: string };
      events: string[];
      delivery_mode: string;
      active: boolean;
    }>;
  }> {
    return this.client.request({
      method: 'GET',
      path: '/messaging/subscriptions',
    });
  }

  /**
   * Poll for events (for polling-mode subscriptions)
   */
  async pollEvents(
    subscriptionId: string,
    options?: { limit?: number; since?: string }
  ): Promise<{
    events: AgentEvent[];
    has_more: boolean;
    next_cursor?: string;
  }> {
    return this.client.request({
      method: 'GET',
      path: `/messaging/subscriptions/${subscriptionId}/poll`,
      query: options,
    });
  }

  /**
   * Stream events using Server-Sent Events (SSE)
   * Returns an async iterator for processing events
   */
  async *streamEvents(options: EventStreamOptions): AsyncIterableIterator<AgentEvent> {
    const url = this.buildStreamUrl(options);
    const response = await fetch(url, {
      headers: this.buildHeaders(),
    });

    if (!response.ok) {
      throw new Error(`Stream failed: ${response.status} ${response.statusText}`);
    }

    const reader = response.body?.getReader();
    if (!reader) {
      throw new Error('Response body is not readable');
    }

    const decoder = new TextDecoder();
    let buffer = '';

    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n\n');
        buffer = lines.pop() || '';

        for (const line of lines) {
          if (line.startsWith('data: ')) {
            const data = line.slice(6);
            if (data === '[DONE]') return;

            try {
              const event = JSON.parse(data) as AgentEvent;
              yield event;
            } catch (e) {
              console.error('Failed to parse event:', e);
            }
          }
        }
      }
    } finally {
      reader.releaseLock();
    }
  }

  /**
   * Build stream URL with query parameters
   */
  private buildStreamUrl(options: EventStreamOptions): string {
    const params = new URLSearchParams();

    if (options.agent) {
      params.append('publisher', options.agent.publisher);
      params.append('name', options.agent.name);
    }

    if (options.event_types) {
      params.append('events', options.event_types.join(','));
    }

    if (options.from_timestamp) {
      params.append('from', options.from_timestamp);
    }

    if (options.batch_size) {
      params.append('batch_size', options.batch_size.toString());
    }

    const baseUrl = (this.client as any).baseUrl;
    return `${baseUrl}/messaging/stream?${params.toString()}`;
  }

  /**
   * Build headers for streaming
   */
  private buildHeaders(): Record<string, string> {
    return (this.client as any).buildHeaders();
  }
}

File: examples/reference-implementations/python-client/ossa_client/messaging.py
"""
OSSA Agent-to-Agent Messaging

Client methods for A2A messaging, webhooks, and event streaming.
"""

from typing import Any, Dict, Iterator, List, Optional
from .client import OSSAClient


class MessagingClient:
    """
    Messaging operations client for A2A communication.
    """

    def __init__(self, client: OSSAClient):
        self._client = client

    def send_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
        """
        Send an A2A message.

        Args:
            message: Message with from, to, type, capability, payload, and metadata

        Returns:
            Message ID and delivery status
        """
        return self._client.request('POST', '/messaging/send', body=message)

    def send_request(self, message: Dict[str, Any], timeout: int = 30000) -> Dict[str, Any]:
        """
        Send a request and wait for response (synchronous-style).

        Args:
            message: Message (without 'type' field)
            timeout: Timeout in milliseconds

        Returns:
            Request ID, response, and status
        """
        return self._client.request(
            'POST',
            '/messaging/request',
            body={**message, 'timeout': timeout},
        )

    def get_message_status(self, message_id: str) -> Dict[str, Any]:
        """
        Get message status.

        Args:
            message_id: Message ID

        Returns:
            Message status and delivery information
        """
        return self._client.request('GET', f'/messaging/messages/{message_id}')

    def broadcast(self, message: Dict[str, Any], filters: Dict[str, Any]) -> Dict[str, Any]:
        """
        Broadcast message to multiple agents.

        Args:
            message: Message to broadcast (without 'to' and 'type' fields)
            filters: Recipient filters (domain, capability, tags)

        Returns:
            Broadcast ID and recipient count
        """
        return self._client.request(
            'POST',
            '/messaging/broadcast',
            body={'message': message, 'filters': filters},
        )

    def register_webhook(self, config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Register a webhook.

        Args:
            config: Webhook configuration (url, events, filters, headers, retry_config)

        Returns:
            Webhook ID and configuration
        """
        return self._client.request('POST', '/messaging/webhooks', body=config)

    def list_webhooks(self) -> Dict[str, Any]:
        """
        List registered webhooks.

        Returns:
            List of webhooks with their configurations
        """
        return self._client.request('GET', '/messaging/webhooks')

    def get_webhook(self, webhook_id: str) -> Dict[str, Any]:
        """
        Get webhook details.

        Args:
            webhook_id: Webhook ID

        Returns:
            Webhook configuration and stats
        """
        return self._client.request('GET', f'/messaging/webhooks/{webhook_id}')

    def update_webhook(self, webhook_id: str, config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Update webhook configuration.

        Args:
            webhook_id: Webhook ID
            config: Updated configuration

        Returns:
            Updated webhook configuration
        """
        return self._client.request('PUT', f'/messaging/webhooks/{webhook_id}', body=config)

    def delete_webhook(self, webhook_id: str) -> None:
        """
        Delete webhook.

        Args:
            webhook_id: Webhook ID
        """
        self._client.request('DELETE', f'/messaging/webhooks/{webhook_id}')

    def subscribe(self, subscription: Dict[str, Any]) -> Dict[str, Any]:
        """
        Subscribe to agent events.

        Args:
            subscription: Subscription config (agent, events, filters, delivery_mode, config)

        Returns:
            Subscription ID and status
        """
        return self._client.request('POST', '/messaging/subscriptions', body=subscription)

    def list_subscriptions(self) -> Dict[str, Any]:
        """
        List event subscriptions.

        Returns:
            List of active subscriptions
        """
        return self._client.request('GET', '/messaging/subscriptions')

    def poll_events(
        self,
        subscription_id: str,
        limit: Optional[int] = None,
        since: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Poll for events (for polling-mode subscriptions).

        Args:
            subscription_id: Subscription ID
            limit: Maximum events to return
            since: Return events since this cursor

        Returns:
            Events list and pagination info
        """
        query = {}
        if limit:
            query['limit'] = limit
        if since:
            query['since'] = since

        return self._client.request('GET', f'/messaging/subscriptions/{subscription_id}/poll', query=query)

    def stream_events(self, options: Optional[Dict[str, Any]] = None) -> Iterator[Dict[str, Any]]:
        """
        Stream events using Server-Sent Events (SSE).

        Args:
            options: Stream options (agent, event_types, from_timestamp, batch_size)

        Yields:
            Agent events as they arrive

        Note:
            This is a basic implementation. For production use, consider using
            a dedicated SSE library like `sseclient-py`.
        """
        import json

        params = {}
        if options:
            if 'agent' in options:
                params['publisher'] = options['agent']['publisher']
                params['name'] = options['agent']['name']
            if 'event_types' in options:
                params['events'] = ','.join(options['event_types'])
            if 'from_timestamp' in options:
                params['from'] = options['from_timestamp']
            if 'batch_size' in options:
                params['batch_size'] = options['batch_size']

        url = self._client._build_url('/messaging/stream', params)
        headers = self._client._build_headers()

        import requests

        response = requests.get(url, headers=headers, stream=True, timeout=None)
        response.raise_for_status()

        buffer = ''
        for line in response.iter_lines(decode_unicode=True):
            if line:
                buffer += line + '\n'
                if line.startswith('data: '):
                    data = line[6:]
                    if data == '[DONE]':
                        return

                    try:
                        event = json.loads(data)
                        yield event
                    except json.JSONDecodeError:
                        continue
